File: frontend/src/services/resume.ts
import { api } from './api'
import type { Resume, ResumeAnalysis } from '@/types'

export const resumeService = {
  async upload(file: File): Promise<Resume> {
    const form = new FormData()
    form.append('file', file)
    const { data } = await api.post<Resume>('/resumes/upload', form, {
      headers: { 'Content-Type': undefined },
    })
    return data
  },

  async list(): Promise<Resume[]> {
    const { data } = await api.get<Resume[]>('/resumes/')
    return data
  },

  async get(id: string): Promise<Resume> {
    const { data } = await api.get<Resume>(`/resumes/${id}`)
    return data
  },

  async remove(id: string): Promise<void> {
    await api.delete(`/resumes/${id}`)
  },

  async analyze(resumeId: string, jobDescription?: string): Promise<ResumeAnalysis> {
    const { data } = await api.post<ResumeAnalysis>(`/resumes/${resumeId}/analyze`, {
      job_description: jobDescription || null,
    })
    return data
  },

  async listAnalyses(resumeId: string): Promise<ResumeAnalysis[]> {
    const { data } = await api.get<ResumeAnalysis[]>(`/resumes/${resumeId}/analyses`)
    return data
  },
}

File: backend/app/routers/resume.py
from fastapi import APIRouter, Depends, UploadFile, File, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from app.database import get_db
from app.models.user import User
from app.models.resume import Resume
from app.schemas.resume import ResumeRead, ResumeAnalysisRead, AnalysisRequest
from app.services import resume_service
from app.routers.deps import get_current_user

router = APIRouter(prefix="/resumes", tags=["resumes"])


@router.post("/upload", response_model=ResumeRead, status_code=201)
async def upload_resume(
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    filename, file_path, file_size = await resume_service.save_resume_file(file, str(current_user.id))
    return resume_service.create_resume(db, current_user, filename, file.filename, file_path, file_size)


@router.get("/", response_model=List[ResumeRead])
def list_resumes(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return db.query(Resume).filter(Resume.owner_id == current_user.id).order_by(Resume.created_at.desc()).all()


@router.get("/{resume_id}", response_model=ResumeRead)
def get_resume(resume_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    resume = db.query(Resume).filter(Resume.id == resume_id, Resume.owner_id == current_user.id).first()
    if not resume:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Resume not found")
    return resume


@router.delete("/{resume_id}", status_code=204)
def delete_resume(resume_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    resume = db.query(Resume).filter(Resume.id == resume_id, Resume.owner_id == current_user.id).first()
    if not resume:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Resume not found")
    db.delete(resume)
    db.commit()


@router.post("/{resume_id}/analyze", response_model=ResumeAnalysisRead)
def analyze_resume(
    resume_id: str,
    request: AnalysisRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    resume = db.query(Resume).filter(Resume.id == resume_id, Resume.owner_id == current_user.id).first()
    if not resume:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Resume not found")
    return resume_service.create_analysis(db, current_user, resume, request.job_description)


@router.get("/{resume_id}/analyses", response_model=List[ResumeAnalysisRead])
def list_analyses(resume_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    resume = db.query(Resume).filter(Resume.id == resume_id, Resume.owner_id == current_user.id).first()
    if not resume:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Resume not found")
    return resume.analyses
